### DISCUSSION TITLE
Rollback when using pytest

### DISCUSSION THREAD

**Discussion by DanielAndreasen** [fastapi/sqlmodel#940] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import os

import pytest
import sqlalchemy as sa
from sqlalchemy.orm import sessionmaker
from sqlmodel import Session

url = f"postgresql+psycopg2://{os.getenv('PGUSER')}:{os.getenv('PGPASSWORD')}@{os.getenv('PGHOST')}/{os.getenv('MYDB')}"
engine = sa.create_engine(url)
TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


@pytest.fixture()
def session():
    connection = engine.connect()
    transaction = connection.begin()
    session = TestingSessionLocal(bind=connection)

    nested = connection.begin_nested()

    @sa.event.listens_for(session, "after_transaction_end")
    def end_savepoint(session, transaction):
        nonlocal nested
        if not nested.is_active:
            nested = connection.begin_nested()

    yield session

    # Run this after the test (regardless of status)
    session.close()
    transaction.rollback()
    connection.close()
```

### Description

I want to test the API I'm writing with fastAPI and SQLModel. The API commits to the database, but I want to make a rollback when testing.
The code attached shows how to do this without problem using sqlalchemy.

The problem is that the `session` from sqlalchemy is slighty different than the one from SQLModel, which I prefer to use, as this is actually the one being used in the API.

If I change the line:
`session = TestingSessionLocal(bind=connection)`
to:
`sesion = Session(engine)  # Session from SQLModel`
then it does not enter the `sa.event.listens_for`, and the rollback never happens.

I'm open for another solution as well.

### Operating System

Linux

### SQLModel Version

0.0.16

### Python Version

Python 3.12.0

### Additional Context

Inspiration found here.
https://stackoverflow.com/questions/67255653/how-to-set-up-and-tear-down-a-database-between-
tests-in-fastapi

**Comment by DanielAndreasen** [comment, marked as answer]
I was able to solve this after many trials and errors (mostly errors).

This is the code:
```python
import pytest
import sqlalchemy as sa
from sqlmodel import Session

from myapp.db.database import engine


@pytest.fixture()
def session():
    connection = engine.connect()
    transaction = connection.begin()
    session = Session(bind=connection)

    nested = connection.begin_nested()

    @sa.event.listens_for(session, "after_transaction_end")
    def end_savepoint(session, transaction):
        nonlocal nested
        if not nested.is_active:
            nested = connection.begin_nested()

    yield session

    session.close()
    transaction.rollback()
    connection.close()
```

The trick was to use `Session(bind=connection)` instead of `Session(engine)`

**Reply by nolzie** [reply]
AWESOME! I was just trying to do the same thing...
